File: 952-word-subsets/word-subsets.java
class Solution {
    public List<String> wordSubsets(String[] words1, String[] words2) {
        int[] cnt = new int[26] ;
        List<String>ans = new ArrayList<>() ;
        // for(int i = 0 ; i < 26 ; i ++) cnt[i] = 0 ;
        for(String s : words2){
            int[] tcnt = new int[26] ;
            for(int i = 0 ; i < s.length() ; i ++){
                tcnt[s.charAt(i) - 'a'] ++ ;
            }
            for(int i = 0 ; i < 26 ; i ++)
                cnt[i] = Math.max(cnt[i], tcnt[i]) ;
        }

        for(String s : words1){
            int[] tcnt = new int[26] ;
            for(int i = 0 ; i < s.length() ; i ++){
                tcnt[s.charAt(i) - 'a'] ++ ;
            }
            boolean ok = true ;
            for(int i = 0 ; i < 26 ; i ++)
                ok &= (tcnt[i] >= cnt[i]) ;
            if(ok) ans.add(s) ;
        }
        return ans ;
    }
}

File: 952-word-subsets/word-subsets.py
class Solution:
    def wordSubsets(self, words1: List[str], words2: List[str]) -> List[str]:
        cnt = [0] * 26 

        def freQ(s):
            tcnt = [0] * 26 
            for i in s:
                tcnt[ord(i) - ord('a')] += 1
            return tcnt 

        for s in words2:
            cnt = [max(i, j) for i, j in zip(cnt, freQ(s))] 
        
        ans = [] 
        for s in words1:
            tcnt = freQ(s)
            if sum(1 for i in range(26) if tcnt[i] >= cnt[i]) == 26:
                ans.append(s)
        
        return ans 
